## Implementation of k-Nearest Neighbors algorithm in Python from scratch

#### Breaking it Down – Pseudo Code of KNN

We can implement a KNN model by following the below steps:

<ol>
    <li>Load the data</li>
    <li>Initialise the value of k</li>
    <li>For getting the predicted class, iterate from 1 to total number of training data points</li>
    <ol>
       <li>Calculate the distance between test data and each row of training data. Here we will use Euclidean distance as our distance metric since it’s the most popular method. The other metrics that can be used are Chebyshev, cosine, etc.</li>
        <li>Sort the calculated distances in ascending order based on distance values</li>
        <li>Get top k rows from the sorted array</li>
        <li>Get the most frequent class of these rows</li>
        <li>Return the predict class</li>
    </ol>
</ol>

#### Implementation in Python from scratch
We will be using the popular _Iris dataset_ for building our KNN model.

In [1]:
# Importing libraries
from sklearn.datasets import load_iris
import pandas as pd
import numpy as np
import math
import operator

In [4]:
#### Start of STEP 1
# Importing data
iris = load_iris()
df = pd.DataFrame(np.c_[iris.data, iris.target], columns=iris.feature_names + ['target'])
#### End of STEP 1

df.head() 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0

In [5]:
# Defining a function which calculates euclidean distance between two data points
def euclideanDistance(obs_1, obs_2, length):
    distance = 0
    for x in range(length):
        distance += np.square(obs_1[x] - obs_2[x])
    return np.sqrt(distance)

# Defining our KNN model
def knn(trainingSet, testInstance, k):
 
    distances = {}
    sort = {}
 
    length = testInstance.shape[1]
    
    #### Start of STEP 3
    # Calculating euclidean distance between each row of training data and test data
    for x in range(len(trainingSet)):
        
        #### Start of STEP 3.A
        dist = euclideanDistance(testInstance, trainingSet.iloc[x], length)

        distances[x] = dist[0]
        #### End of STEP 3.A
 
    #### Start of STEP 3.B
    # Sorting them on the basis of distance
    sorted_d = sorted(distances.items(), key=operator.itemgetter(1))
    #### End of STEP 3.B
 
    neighbors = []
    
    #### Start of STEP 3.C
    # Extracting top k neighbors
    for x in range(k):
        neighbors.append(sorted_d[x][0])
    #### End of STEP 3.C
    classVotes = {}
    
    #### Start of STEP 3.D
    # Calculating the most freq class in the neighbors
    for x in range(len(neighbors)):
        response = trainingSet.iloc[neighbors[x]][-1]
 
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    #### End of STEP 3.D

    #### Start of STEP 3.E
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    return(sortedVotes[0][0], neighbors)
    #### End of STEP 3.E

#### Application of k-NN algorithm on _Iris dataset_

In [10]:
# Creating a dummy testset
testSet = [[7.2, 3.6, 5.1, 2.5]]
test = pd.DataFrame(testSet)

#### Start of STEP 2
# Setting number of neighbors = 1
k = 1
#### End of STEP 2
# Running KNN model
result,neigh = knn(df, test, k)

# Predicted class
print('Predicted class for the given vector (k=1) is {}'.format(result))

# Nearest neighbor
print('Nearest neighbor for the predicted class is {}'.format(neigh))

Predicted class for the given vector (k=1) is 2.0
Nearest neighbor for the predicted class is [141]


Now we will try to alter the k values, and see how the prediction changes.

In [14]:
# Setting number of neighbors = 3 
k = 3 
# Running KNN model 
result,neigh = knn(df, test, k) 
# Predicted class 
print('Predicted class for the given vector (k=3) is {}'.format(result))
# 3 nearest neighbors
print('Nearest neighbor for the predicted class is {}'.format(neigh))

print()
# Setting number of neighbors = 5
k = 5
# Running KNN model 
result,neigh = knn(df, test, k) 
# Predicted class 
print('Predicted class for the given vector (k=5) is {}'.format(result))
# 5 nearest neighbors
print('Nearest neighbor for the predicted class is {}'.format(neigh))

Predicted class for the given vector (k=3) is 2.0
Nearest neighbor for the predicted class is [141, 139, 120]

Predicted class for the given vector (k=5) is 2.0
Nearest neighbor for the predicted class is [141, 139, 120, 145, 144]


#### Comparing our model with scikit-learn

In [16]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(df.iloc[:,0:4], df['target'])

# Predicted class
print('Predicted class for the given vector (k=5) is {}'.format(neigh.predict(test)))

# 3 nearest neighbors
print('Nearest neighbor for the predicted class is {}'.format(neigh.kneighbors(test)[1]))

Predicted class for the given vector (k=5) is [2.]
Nearest neighbor for the predicted class is [[141 139 120]]


#### Conclusion
We can see that both the models predicted the same class (‘Iris-virginica’) and the same nearest neighbors ( [141 139 120] ). Hence we can conclude that our model runs as expected.